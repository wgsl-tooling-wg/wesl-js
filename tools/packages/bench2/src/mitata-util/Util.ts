/** return an array partitioned into possibly overlapping groups */
export function grouped<T>(a: T[], size: number, stride = size): T[][] {
  const groups = [];
  for (let i = 0; i < a.length; i += stride) {
    groups.push(a.slice(i, i + size));
  }
  return groups;
}

/** apply a map() on the values in a Record.
 * @return a new Record.  */
export function mapValues<K extends string | number | symbol, T, U>(
  obj: Record<K, T>,
  fn: (v: T) => U,
): Record<K, U> {
  const result = {} as Record<K, U>;
  for (const key in obj) {
    if (Object.hasOwn(obj, key)) {
      result[key] = fn(obj[key]);
    }
  }
  return result;
}
