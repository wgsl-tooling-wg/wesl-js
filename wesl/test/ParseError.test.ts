import { expect, test } from "vitest";
import { errorHighlight } from "../Util.ts";
import { parseTest } from "./TestUtil.ts";
import { assertSnapshot } from "@std/testing/snapshot";

function getError(fn: () => void) {
  try {
    fn();
  } catch (e) {
    return e;
  }
}

test("parse fn foo() { invalid }", async (t) => {
  const src = "fn foo() { let }";
  await assertSnapshot(t, getError(() => parseTest(src)));
});

test("parse invalid if", async (t) => {
  const src = `fn foo() { 
  let a = 3;
  if(1<1) { 🐈‍⬛ } else {  }
  }`;
  await assertSnapshot(t, getError(() => parseTest(src)));
});

test("parse invalid name", async (t) => {
  const src = "var package = 3;";
  await assertSnapshot(t, getError(() => parseTest(src)));
});

test("error highlight", () => {
  expect(errorHighlight("foo", [0, 2]).join("\n")).toBe(`foo
^^`);
  expect(errorHighlight("foo", [0, 1]).join("\n")).toBe(`foo
^`);
  expect(errorHighlight("foo", [0, 0]).join("\n")).toBe(`foo
^`);
  expect(errorHighlight("foo", [1, 2]).join("\n")).toBe(`foo
 ^`);
});
