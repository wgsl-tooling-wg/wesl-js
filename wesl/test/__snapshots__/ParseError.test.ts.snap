export const snapshot = {};

snapshot[`parse fn foo() { invalid } 1`] = `
Error: ./test.wesl:1:15 error: invalid ident
fn foo() { let }
              ^^
    at eval (./test.wesl, <anonymous>:1:27)
Caused by Error
    at parseSrcModule (file:///E:/git/webgpu/wesl-js/wesl/ParseWESL.ts:106:21)
    at parseWESL (file:///E:/git/webgpu/wesl-js/wesl/test/TestUtil.ts:32:10)
    at parseTestRaw (file:///E:/git/webgpu/wesl-js/wesl/test/TestUtil.ts:112:10)
    at file:///E:/git/webgpu/wesl-js/wesl/test/TestUtil.ts:107:28
    at withLogger (file:///E:/git/webgpu/wesl-js/mini-parse/WrappedLog.ts:9:12)
    at expectNoLog (file:///E:/git/webgpu/wesl-js/mini-parse/test-util/TestParse.ts:78:14)
    at parseTest (file:///E:/git/webgpu/wesl-js/wesl/test/TestUtil.ts:107:10)
    at file:///E:/git/webgpu/wesl-js/wesl/test/ParseError.test.ts:16:42
    at getError (file:///E:/git/webgpu/wesl-js/wesl/test/ParseError.test.ts:8:5)
    at file:///E:/git/webgpu/wesl-js/wesl/test/ParseError.test.ts:16:27
Caused by Error: invalid ident
    at Parser._req [as fn] (file:///E:/git/webgpu/wesl-js/mini-parse/ParserCombinator.ts:520:13)
    at Parser._run (file:///E:/git/webgpu/wesl-js/mini-parse/Parser.ts:161:27)
    at Parser._seq [as fn] (file:///E:/git/webgpu/wesl-js/mini-parse/ParserCombinator.ts:185:24)
    at Parser._run (file:///E:/git/webgpu/wesl-js/mini-parse/Parser.ts:161:27)
    at Parser._or [as fn] (file:///E:/git/webgpu/wesl-js/mini-parse/ParserCombinator.ts:329:24)
    at Parser._run (file:///E:/git/webgpu/wesl-js/mini-parse/Parser.ts:161:27)
    at runAndCollectAfter (file:///E:/git/webgpu/wesl-js/mini-parse/ParserCollect.ts:164:20)
    at Parser._tagScope [as fn] (file:///E:/git/webgpu/wesl-js/mini-parse/ParserCollect.ts:118:14)
    at Parser._run (file:///E:/git/webgpu/wesl-js/mini-parse/Parser.ts:161:27)
    at Parser._fn (file:///E:/git/webgpu/wesl-js/mini-parse/ParserCombinator.ts:643:21)
`;

snapshot[`parse invalid if 1`] = `
Error: ./test.wesl:3:13 error: Invalid token 🐈

  if(1<1) { 🐈‍⬛ } else {  }
             ^^
    at eval (./test.wesl, <anonymous>:1:27)
Caused by Error
    at parseSrcModule (file:///E:/git/webgpu/wesl-js/wesl/ParseWESL.ts:106:21)
    at parseWESL (file:///E:/git/webgpu/wesl-js/wesl/test/TestUtil.ts:32:10)
    at parseTestRaw (file:///E:/git/webgpu/wesl-js/wesl/test/TestUtil.ts:112:10)
    at file:///E:/git/webgpu/wesl-js/wesl/test/TestUtil.ts:107:28
    at withLogger (file:///E:/git/webgpu/wesl-js/mini-parse/WrappedLog.ts:9:12)
    at expectNoLog (file:///E:/git/webgpu/wesl-js/mini-parse/test-util/TestParse.ts:78:14)
    at parseTest (file:///E:/git/webgpu/wesl-js/wesl/test/TestUtil.ts:107:10)
    at file:///E:/git/webgpu/wesl-js/wesl/test/ParseError.test.ts:24:42
    at getError (file:///E:/git/webgpu/wesl-js/wesl/test/ParseError.test.ts:8:5)
    at file:///E:/git/webgpu/wesl-js/wesl/test/ParseError.test.ts:24:27
Caused by Error: Invalid token 🐈
    at WeslStream.nextToken (file:///E:/git/webgpu/wesl-js/wesl/parse/WeslStream.ts:119:15)
    at skipIgnored (file:///E:/git/webgpu/wesl-js/mini-parse/ParserCollect.ts:293:25)
    at refinePosition (file:///E:/git/webgpu/wesl-js/mini-parse/ParserCollect.ts:287:17)
    at queueCollectFn (file:///E:/git/webgpu/wesl-js/mini-parse/ParserCollect.ts:178:23)
    at Parser._tagScope [as fn] (file:///E:/git/webgpu/wesl-js/mini-parse/ParserCollect.ts:109:7)
    at Parser._run (file:///E:/git/webgpu/wesl-js/mini-parse/Parser.ts:161:27)
    at Parser._or [as fn] (file:///E:/git/webgpu/wesl-js/mini-parse/ParserCombinator.ts:329:24)
    at Parser._run (file:///E:/git/webgpu/wesl-js/mini-parse/Parser.ts:161:27)
    at Parser._fn (file:///E:/git/webgpu/wesl-js/mini-parse/ParserCombinator.ts:643:21)
    at Parser._run (file:///E:/git/webgpu/wesl-js/mini-parse/Parser.ts:161:27)
`;

snapshot[`parse invalid name 1`] = `
Error: ./test.wesl:1:4 error: expected identifier
var package = 3;
   ^^^^^^^^
    at eval (./test.wesl, <anonymous>:1:27)
Caused by Error
    at parseSrcModule (file:///E:/git/webgpu/wesl-js/wesl/ParseWESL.ts:106:21)
    at parseWESL (file:///E:/git/webgpu/wesl-js/wesl/test/TestUtil.ts:32:10)
    at parseTestRaw (file:///E:/git/webgpu/wesl-js/wesl/test/TestUtil.ts:112:10)
    at file:///E:/git/webgpu/wesl-js/wesl/test/TestUtil.ts:107:28
    at withLogger (file:///E:/git/webgpu/wesl-js/mini-parse/WrappedLog.ts:9:12)
    at expectNoLog (file:///E:/git/webgpu/wesl-js/mini-parse/test-util/TestParse.ts:78:14)
    at parseTest (file:///E:/git/webgpu/wesl-js/wesl/test/TestUtil.ts:107:10)
    at file:///E:/git/webgpu/wesl-js/wesl/test/ParseError.test.ts:29:42
    at getError (file:///E:/git/webgpu/wesl-js/wesl/test/ParseError.test.ts:8:5)
    at file:///E:/git/webgpu/wesl-js/wesl/test/ParseError.test.ts:29:27
Caused by Error: expected identifier
    at Parser._req [as fn] (file:///E:/git/webgpu/wesl-js/mini-parse/ParserCombinator.ts:520:13)
    at Parser._run (file:///E:/git/webgpu/wesl-js/mini-parse/Parser.ts:161:27)
    at runAndCollectAfter (file:///E:/git/webgpu/wesl-js/mini-parse/ParserCollect.ts:164:20)
    at Parser._tagScope [as fn] (file:///E:/git/webgpu/wesl-js/mini-parse/ParserCollect.ts:118:14)
    at Parser._run (file:///E:/git/webgpu/wesl-js/mini-parse/Parser.ts:161:27)
    at runAndCollectAfter (file:///E:/git/webgpu/wesl-js/mini-parse/ParserCollect.ts:164:20)
    at Parser._ctag [as fn] (file:///E:/git/webgpu/wesl-js/mini-parse/ParserCollect.ts:142:12)
    at Parser._run (file:///E:/git/webgpu/wesl-js/mini-parse/Parser.ts:161:27)
    at Parser._seq [as fn] (file:///E:/git/webgpu/wesl-js/mini-parse/ParserCombinator.ts:185:24)
    at Parser._run (file:///E:/git/webgpu/wesl-js/mini-parse/Parser.ts:161:27)
`;
